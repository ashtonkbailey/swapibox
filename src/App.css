@import url('https://fonts.googleapis.com/css?family=Poller+One');

.App {
  text-align: center;
  background-image: url('./images/starwars.jpg');
  background-size: cover;
  height: 100vh;
  padding: 0;
  margin: 0;
}


.header {
  font-family: 'Poller One', cursive;
  margin-top: 0;
  margin-bottom: 0;
  padding: 40px;
  text-shadow:
		-3px -3px 0 rgb(244, 217, 68),
		2px -2px 0 rgb(244, 217, 68),
		-2px 2px 0 rgb(244, 217, 68),
		2px 2px 0 rgb(244, 217, 68);
  font-size: 88px;
  text-align: center;
  color:black;
  border: 4px solid #000000;

}
