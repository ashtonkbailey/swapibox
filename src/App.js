import React, { Component } from 'react';
import './App.css';
import Navigation from './components/navigation/Navigation';
import Favorites from './components/favorites/Favorites';
import ContentContainer from './components/contentContainer/ContentContainer';
import People from './components/Helpers/People';
import Planets from './components/Helpers/Planets';
import Vehicles from './components/Helpers/Vehicles';

class App extends Component {
  constructor() {
    super();
    this.state = {
      errorStatus: '',
      favorites: [],
      displayedContent: [],
      chosenContent: '',
      currFilm: {
        title: '',
        crawl: '',
        year: '',
      },
      carouselIndex: 0,
    };
  }

  cleanFilm = (film) => {
    const cleanCrawl = film.opening_crawl.replace(/(\r\n|\n|\r)/gm, ' ');
    const cleanYear = film.release_date.substring(0, 4);
    return {
      title: film.title,
      crawl: cleanCrawl,
      year: cleanYear,
    };
  }

  componentDidMount = async () => {
    const randomNum = this.generateRandomNum();
    try {
      const response = await fetch(`https://swapi.co/api/films/${randomNum}`);
      const film = await response.json();
      const cleanFilm = await this.cleanFilm(film);
      this.setState({
        currFilm: {
          title: cleanFilm.title,
          crawl: cleanFilm.crawl,
          year: cleanFilm.year,
        },
      });
    } catch (error) {
      this.setState({
        errorStatus: `Error: ${error.message}`,
      });
    }
    localStorage.setItem('current film', JSON.stringify(this.state.currFilm));
  }

  displayChosenContent = (e) => {
    const contentToFetch = e.target.name;
    this.fetchChosenContent(contentToFetch);
  }

  fetchPropertyObj = async (url) => {
    const response = await fetch(url);
    const propertyObj = await response.json();
    return propertyObj;
  }

  displayData = (name) => {
    if (Object.keys(localStorage).includes(name)) {
      this.getDataFromStorage(name);
    } else {
      this.fetchChosenContent(name);
    }
  }

  getDataFromStorage = (name) => {
    const data = JSON.parse(localStorage.getItem(name));
  }

  fetchChosenContent = async (name) => {
    let cleanedData;
    if (name === 'people') {
      cleanedData = await new People();
    } else if (name === 'planets') {
      cleanedData = await new Planets();
    } else if (name === 'vehicles') {
      cleanedData = await new Vehicles();
    }
    localStorage.setItem(`${name}`, JSON.stringify([...cleanedData]));
    await this.setState({
      displayedContent: cleanedData,
      chosenContent: name,
    });
  }

  incrementCarousel = () => {
    const { carouselIndex } = this.state;
    let newIndex = 0;

    if (carouselIndex < 9) {
      newIndex = carouselIndex + 1;
    }

    this.setState({ carouselIndex: newIndex });
  }

  decrementCarousel = () => {
    const { carouselIndex } = this.state;
    let newIndex;

    if (carouselIndex < 1) {
      newIndex = 9;
    } else {
      newIndex = carouselIndex - 1;
    }

    this.setState({ carouselIndex: newIndex });
  }

  viewFavorites = () => {
    console.log(this.state.favorites);
  }

  addToFavorites = (favoritedCard) => {
    const clickedCard = favoritedCard;
    clickedCard.card.favorite = true;
    const { favorites } = this.state;

    if (!favorites.includes(clickedCard.card)) {
      this.setState({
        favorites: [...favorites, clickedCard.card],
      });
    }
  }

  generateRandomNum = () => Math.floor(Math.random() * 6) + 1;

  render() {
    const {
      favorites,
      chosenContent,
      currFilm,
      displayedContent,
      carouselIndex,
    } = this.state;

    return (
      <div className="App">
        <h1 className="header"> SWAPIBOX </h1>
        <Navigation displayChosenContent={this.displayChosenContent} />
        <Favorites
          faves={favorites}
          viewFavorites={this.viewFavorites}
        />
        <ContentContainer
          chosenContent={chosenContent}
          film={currFilm}
          contents={displayedContent}
          incrementCarousel={this.incrementCarousel}
          decrementCarousel={this.decrementCarousel}
          carouselIndex={carouselIndex}
          addToFavorites={this.addToFavorites}
        />
      </div>
    );
  }
}

export default App;
