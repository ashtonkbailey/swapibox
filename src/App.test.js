import React from 'react';
import { shallow } from 'enzyme';
import App from './App';

describe('App', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<App />);
  });

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('cleanFilm', () => {
    it('should clean up film when cleanFilm is called', () => {
      const film = {
        title: 'A New Hope',
        opening_crawl: 'It is a period of civil war. Rebel spaceships, striking from a hidden base, have won their first victory against the evil Galactic Empire.',
        release_date: '1977-05-25',
      };
      const result = wrapper.instance().cleanFilm(film);
      const expected = {
        title: 'A New Hope',
        crawl: 'It is a period of civil war. Rebel spaceships, striking from a hidden base, have won their first victory against the evil Galactic Empire.',
        year: '1977',
      };

      expect(result).toEqual(expected);
    });
  });

  describe('generateRandomNum', () => {
    it('should generate a random number between 1 and 7', () => {
      // Setup
      Math.random = jest.fn().mockImplementation(() => 0.4);
      const expected = 3;
      // Execution
      const result = wrapper.instance().generateRandomNum();
      // Expectation
      expect(expected).toEqual(result);
    });
  });

  describe('fetchCurrFilm', () => {
    let generateRandomNum;
    let mockCurrFilm;

    beforeEach(() => {
      Math.random = jest.fn().mockImplementation(() => 0.4);
      mockCurrFilm = {
        title: '',
        crawl: '',
        year: '',
      };
      window.fetch = jest.fn().mockImplementation(() => Promise.resolve({
        ok: true,
        json: () => Promise.resolve(mockCurrFilm),
      }));
    });

    it('calls generateRandomNum', () => {
      // Setup
      const mockGenerateRandomNum = jest.fn();
      wrapper.instance().generateRandomNum = mockGenerateRandomNum;
      // Execution
      wrapper.instance().fetchCurrFilm();
      // Expectation
      expect(mockGenerateRandomNum).toHaveBeenCalled();
    });

    it('should call cleanFilm', async () => {
      // Setup
      const mockCleanFilm = jest.fn();
      wrapper.instance().cleanFilm = mockCleanFilm;
      // Execution
      await wrapper.instance().fetchCurrFilm();
      // Expectation
      expect(mockCleanFilm).toHaveBeenCalled();
    });

    it('calls fetch with the correct parameters', () => {
      // Setup
      const expected = 'https://swapi.co/api/films/3';
      // Execution
      wrapper.instance().fetchCurrFilm();
      // Expectation
      expect(window.fetch).toHaveBeenCalledWith(expected);
    });

    it('should setState', async () => {
      // Setup
      const mockCleanFilm = jest.fn(() => ({ title: 'blahtitle', crawl: 'blahcrawl', year: 1770 }));
      const expected = { title: 'blahtitle', crawl: 'blahcrawl', year: 1770 };
      wrapper.instance().cleanFilm = mockCleanFilm;
      // Execution
      await wrapper.instance().fetchCurrFilm();
      // Expectation
      expect(wrapper.state().currFilm).toEqual(expected);
    });

    it('on fetch error, errorStatus is set', async () => {
      // Setup
      window.fetch = jest.fn().mockImplementation(() => Promise.reject({
        message: 'mock failure',
      }));
      // Execution
      await wrapper.instance().fetchCurrFilm();
      // Expectation
      expect(wrapper.state().errorStatus).toEqual('Error: mock failure');
    });
  });

  describe('displayChosenContent', () => {
    it.skip('should call fetchChosenContent', () => {
      // Setup
      const mockEvent = { preventDefault: () => {} };
      const mockFetchChosenContent = jest.fn();
      // Execution
      wrapper.instance().displayChosenContent(mockEvent);
      // Expectation
      expect(mockFetchChosenContent).toHaveBeenCalled();
    });
  });

  describe('fetchPropertyObj', () => {
    it('calls fetch with the correct parameter', async () => {
      // Setup

      // Execution

      // Expectation

    });

    it('returns the property object', async () => {
      // Setup
      
      // Execution

      // Expectation

    });
  });

  describe('cleanPeopleData', () => {
    it('returns a Promise array', () => {
      // Setup

      // Execution

      // Expectation

    });

    it('calls fetchPropertyObj', async () => {
      // Setup

      // Execution

      // Expectation

    });
  });

  describe('fetchChosenContent', () => {
    let mockCleanedPeople;

    beforeEach(() => {
      mockCleanedPeople = [];
      window.fetch = jest.fn().mockImplementation(() => Promise.resolve({
        ok: true,
        json: () => Promise.resolve(mockCleanedPeople),
      }));
    });

    it('calls fetch with the correct parameters', async () => {
      // Setup
      const expected = 'https://swapi.co/api/people/';
      // Execution
      await wrapper.instance().fetchPropertyObj(expected);
      // Expectation
      expect(window.fetch).toHaveBeenCalledWith(expected);
    });

    it('calls cleanPeopleData', async () => {
      // Setup

      // Execution

      // Expectation

    });

    it('updates state', async () => {
      // Setup

      // Execution

      // Expectation

    });
  });

  describe('incrementCarousel', () => {
    it('updates state', () => {
      // Setup

      // Execution

      // Expectation

    });
  });

  describe('decrementCarousel', () => {
    it('updates state', () => {
      // Setup

      // Execution

      // Expectation

    });
  });
  
});
