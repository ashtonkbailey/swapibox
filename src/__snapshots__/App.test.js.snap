// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h1>
           SWAPIBOX 
        </h1>,
        <Navigation />,
        <Favorites
          faves={Array []}
        />,
        <ContentContainer
          film={
            Object {
              "crawl": "",
              "title": "",
              "year": "",
            }
          }
        />,
      ],
      "className": "App",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": " SWAPIBOX ",
        },
        "ref": null,
        "rendered": " SWAPIBOX ",
        "type": "h1",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "faves": Array [],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "film": Object {
            "crawl": "",
            "title": "",
            "year": "",
          },
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h1>
             SWAPIBOX 
          </h1>,
          <Navigation />,
          <Favorites
            faves={Array []}
          />,
          <ContentContainer
            film={
              Object {
                "crawl": "",
                "title": "",
                "year": "",
              }
            }
          />,
        ],
        "className": "App",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": " SWAPIBOX ",
          },
          "ref": null,
          "rendered": " SWAPIBOX ",
          "type": "h1",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "faves": Array [],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "film": Object {
              "crawl": "",
              "title": "",
              "year": "",
            },
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
